Show parameter documentation for delegates in Signature Help

jhinder opened the pull request:
This is the second PR for #26713.

With this PR, Signature Help will show documentation for delegate parameters.

Previously `ISymbolExtensions2.GetDocumentation` would pick the documentation tags of the symbol containing the parameter. However, for delegates this symbol would be the `Invoke` method, which does not have documentation.  
Now we check whether the symbol containing the parameter's containing symbol (i.e. the parameter's "grandparent" symbol) is a delegate, and if so, take the documentation of that symbol.  
The behaviour for all other parameters is not changed.

CyrusNajmabadi commented on src/Features/Core/Portable/Shared/Extensions/ISymbolExtensions_2.cs:
```diff
@@ -182,14 +182,27 @@ private static string GetDocumentation(ISymbol symbol, CancellationToken cancell
+            if (containingSymbol.ContainingSymbol.IsDelegateType())
+            {
+                // containingSymbol is the Invoke method of the delegate, so we need to go one level up and take the method's containing symbol.
+                containingSymbol = containingSymbol.ContainingSymbol;
```
Are we sure th is currect?  How do you know it's the Invoke mehtod?  What if it's hte BeginInvoke method (or any other number of methods that are on delegates)?  can you actually validate that the containing symbol is the 'Invoke' method?

sharwell commented on src/EditorFeatures/CSharpTest/SignatureHelp/InvocationExpressionSignatureHelpProviderTests.cs:
```diff
@@ -205,6 +205,28 @@ void Goo(int a, int b)
+
+    void M(SomeDelegate theDelegate)
+    {
+        [|theDelegate($$|]);
```
❗️ I would expect to see tests for the following forms as well:

* `theDelegate.Invoke($$);`
* `theDelegate.BeginInvoke($$);`
* `theDelegate.BeginInvoke(0, $$);` &larr; ensures that the implicit `AsyncCallback` parameter doesn't cause a problem
* `theDelegate.EndInvoke($$);` &larr; ensures that the `IAsyncResult` parameter is handled correctly, and specifically covers the case described separately by @CyrusNajmabadi

jhinder replied:
In the revised PR, EndInvoke no longer receives "special treatment" (only Invoke/BeginInvoke do, see comment), it is treated like any other method (w.r.t. Signature Help/parameter documentation). Should I still add a unit test for it?

jhinder commented:
I've reworked the code a bit, and added more tests.

Now we check whether a) the grandparent symbol is a delegate and b) the parent symbol is a method symbol.
If that is the case, we check the name of the symbol.  
If the name is "BeginInvoke", and the parameter's position/ordinal is among the last two method parameters (the two implicit params), we return null (no documentation).  
If the name is either "Invoke" or "BeginInvoke" (the two delegate invocation functions whose parameters correspond to the delegate params), we proceed as in the first version of the PR.

sharwell commented:
@dotnet-bot retest windows_debug_spanish_unit32_prtest please

sharwell commented:
@jinujoseph for approval

jinujoseph commented:
approved to merge for 15.8.preview3
